### ISSUE TITLE
Experiment updating thin-edge from Cumulocity 

### ISSUE THREAD

**Issue by didier-wenzek** [thin-edge/thin-edge.io#1466]
One should be able to update thin-edge from the cloud avoiding any manual intervention over all the devices of a fleet.
Can this be done using the software management feature of thin-edge itself?

Unfortunately, I foresee some issues. Mainly because of:
* the current `apt` packaging of thin-edge (there are several packages with inter-dependencies),
* the fact that the mapper and the agent are restarted during the update.

But is this enough to put all the thin-edge executables in a single package with a new approach for restart?
This is the purpose of this task: have a better insight of what must be done to update the thin-edge Debian packages from the cloud.

* Confirm that updating the `tedge` cli works with no issue.
* What is the behavior of the `mapper` update?
   * Is the mapper properly updated and restarted?
   * Is the operation status correctly reported?
* What is the behavior of the `agent` update?
   * Is the update request properly executed to completion?
   * Is the agent properly updated and restarted?
   * Is the operation status correctly reported?

AC:
* Description of the scenarios explored with the expected outcome
* Description of the actual outcome and issues
* Workaround proposals if any

**Comment by rina23q** [comment]
First, we should distinguish distinguish all packages whether "daemon" or "one-time process".

1) One-time process: tedge, tedge_apt_plugin, tedge_apama_plugin
2) Daemon: tedge_mapper, tedge_agent, tedge_watchdog, c8y_configuration_plugin, c8y_log_plugin

For 1), package upgrading via Cumulocity is already feasible without modifying anything. => Out of scope of this issue.
For 2), all have issues on upgrading via Cumulocity. We need further distinguish, a) tedge_agent or b) else. => Target of this issue.

For a), if tedge_agent process stops while upgrading, this interrupts dkpg, and the tedge_agent package goes to broken state. Therefore, I tried these steps as workaround. And it worked with this workaround. :smiley: 
1. While upgrading is on-going, tedge_agent process doesn't stop or restart. (keep running the older version which is already installed)
2. To apply the change, user must call "Restart device" from Cumulocity.
3. After the restart of the device, tedge_agent starts running with the new version.

For b), stop the corresponding process before upgrading and start the process after upgrading is required. This is unfortunately not done correctly in our current maintainers scripts. After changing `prerm` and `postinst` scripts, the self-update works.

**Comment by rina23q** [comment]
# The final version of maintainers scripts
## tedge_agent
### prerm
Removed all contents.
### postrm
Removed the automatically added by cargo-deb (which stops the tedge-agent service while upgrading)
```shell
#!/bin/sh
set -e

purge_agent_directory() {
    if [ -d "/etc/tedge/.agent" ]; then
        rm -rf /etc/tedge/.agent
    fi
}

purge_agent_lock() {
   if [ -f "/run/lock/tedge_agent.lock" ]; then
       rm -rf /run/lock/tedge_agent.lock
   fi
}

case "$1" in
    purge)
       purge_agent_directory
       purge_agent_lock
    ;;

    remove|upgrade|failed-upgrade|abort-install|abort-upgrade|disappear)
    ;;

    *)
        echo "tedge postrm called with unknown argument \`$1\`" >&2
        exit 1
    ;;
esac
```

### postinst
```shell
#!/bin/sh

#set -e

# Initialize the agent
if command -v systemctl >/dev/null; then
    if ! systemctl is-active --quiet tedge-agent; then
         sudo -u tedge -- tedge_agent --init
    fi
fi

# Reenable the services only if systemctl is available
if command -v systemctl >/dev/null; then
    ### Enable the sm services if the device is connected to c8y cloud
    if [ -f "/etc/tedge/mosquitto-conf/c8y-bridge.conf" ] && ! systemctl is-active --quiet tedge-agent ; then
        # start and enable tedge-agent
        systemctl start tedge-agent.service
        systemctl enable tedge-agent.service
    fi
fi
```

**Comment by rina23q** [comment]
To enable self-update of thin-edge components, we need to change maintainers files.

- [ ] Update maintainers files of tedge_agent to enable self-update via APT
- [ ] Update maintainers files of tedge_mapper, tedge_watchdog to enable self-update via APT
- [ ] Update maintainers files of c8y_configuration_plugin to enable self-update via APT

These must be part of 0.8.0 release.

**Comment by rina23q** [comment]
By the way, I got an error when I tried downgrade by `tedge_apt_plugin`. Is it intended?

```
E: Packages were downgraded and -y was used without --allow-downgrades.
EOF
```

**Comment by rina23q** [comment]
> --allow-downgrades

Opened a PR #1524 to fix it.

**Comment by rina23q** [comment]
### Tests for `tedge_mapper c8y`
#### OTA via Cumulocity
Prerequisites:
* DTU has already installed the released version `0.8.0`
* DTU is connected to c8y. The service under test is running.
e.g. Returns 0 by `systemctl is-active --quiet tedge-mapper-c8y`
* The debian packages with version `0.8.0` and `0.8.1` are uploaded in c8y.

Test:
1. Upgrade from `0.8.0` to `0.8.1` via c8y. => 
* Successful on c8y
* `0.8.1` is visible by `apt list --installed | grep tedge_mapper`

**Comment by rina23q** [comment]
### Tests for `tedge_agent`
#### OTA via Cumulocity
Prerequisites:
* DTU has already installed the released version `0.8.0`
* DTU is connected to c8y. The service under test is running.
e.g. Returns 0 by `systemctl is-active --quiet tedge-agent`
* The debian packages with version `0.8.0` and `0.8.1` are uploaded in c8y.

Test:
1. Upgrade from `0.8.0` to `0.8.1` via c8y. => 
* Successful on c8y
* `0.8.1` is visible by `apt list --installed | grep tedge_agent`
2. Trigger "device restart" from c8y. =>
* Successful on c8y
3. Check if `tedge-agent` service is running. Returns 0 by `systemctl is-active --quiet tedge-agent`

**Comment by rina23q** [comment]
Delivered in 0.8.0
